vsphere/internal/virtualdevice/virtual_machine_disk_subresource.go

```diff
@@ -0,0 +1,1514 @@
+// set.
+//
+// The following validation operations are also carried out on the set as a
+// whole:
+//
+// * Ensuring all names are unique across the set.
+// * Ensuring that at least one element in the set has a unit_number of 0.
+func DiskDiffOperation(d *schema.ResourceDiff, c *govmomi.Client) error {
+	log.Printf("[DEBUG] DiskDiffOperation: Beginning disk diff customization")
+	o, n := d.GetChange(subresourceTypeDisk)
+	// Some global validation first. We handle individual validation later.
+	log.Printf("[DEBUG] DiskDiffOperation: Beginning diff validation (indexes aligned to new config)")
+	names := make(map[string]struct{})
+	units := make(map[int]struct{})
+	if len(n.([]interface{})) < 1 {
+		return errors.New("there must be at least one disk specified")
+	}
+	for ni, ne := range n.([]interface{}) {
+		nm := ne.(map[string]interface{})
+		// Because we support long and short-form paths, we don't support duplicate
+		// file names right now, even if they are in a different path. This makes
+		// things slightly more inflexible but hopefully not enough to be seriously
+		// cumbersome to the use of the resource.
+		name := path.Base(nm["name"].(string))
+		if _, ok := names[name]; ok {
+			return fmt.Errorf("disk: duplicate name %s", name)
+		}
+		if _, ok := units[nm["unit_number"].(int)]; ok {
+			return fmt.Errorf("disk: duplicate unit_number %d", nm["unit_number"].(int))
+		}
+		names[name] = struct{}{}
+		units[nm["unit_number"].(int)] = struct{}{}
+		// Run the resource through an individual validate function. This performs
+		// field validation for things we don't need to know the state of other
+		// resources for.
+		r := NewDiskSubresource(c, nil, d, nm, nil, ni)
+		if err := r.ValidateDiff(); err != nil {
+			return fmt.Errorf("%s: %s", r.Addr(), err)
+		}
+	}
+	if _, ok := units[0]; !ok {
+		return errors.New("at least one disk must have a unit_number of 0")
+	}
+
+	// Perform the normalization here.
+	log.Printf("[DEBUG] DiskDiffOperation: Beginning diff normalization (indexes aligned to old state)")
+	ods := o.([]interface{})
+	nds := n.([]interface{})
+
+	normalized := make([]interface{}, len(ods))
+nextNew:
+	for _, ne := range nds {
+		nm := ne.(map[string]interface{})
+		for oi, oe := range ods {
+			om := oe.(map[string]interface{})
+			// We extrapolate using the name as a "primary key" of sorts. Since we
+			// support both long-form and short-form paths, and don't support using
+			// same file name regardless of if you are using a long-from path, we
+			// just check the short-form and do the comparison from there.
+			if path.Base(nm["name"].(string)) == path.Base(om["name"].(string)) {
+				r := NewDiskSubresource(c, nil, d, nm, om, oi)
+				if err := r.NormalizeDiff(); err != nil {
+					return fmt.Errorf("%s: %s", r.Addr(), err)
+				}
+				normalized[oi] = r.Data()
+				continue nextNew
+			}
+		}
+		// We didn't find a match for this resource, it could be a new resource or
```

Comment:
What we are doing here, is guarding against this scenario:

Old:

```
disk {
  name = "system.vmdk"
  size = 20
}

disk {
  name        = "extra1.vmdk"
  unit_number = 1
  size        = 10
}

disk {
  name        = "extra2.vmdk"
  unit_number = 2
  size        = 5
}
```

New:

```
disk {
  name = "system.vmdk"
  size = 20
}

disk {
  name        = "extra3.vmdk"
  unit_number = 1
  size        = 10
}

disk {
  name        = "extra2.vmdk"
  unit_number = 2
  size        = 5
}
```

In this scenario, the old middle disk (`extra1.vmdk`) has been removed, and its slot is now being taken up by a new disk (`extra3.vmdk`). As this is a list, the state will still have the data from the old disk, which is incorrect - this is a new device and should not have a key or device address yet. The rest of the data is just fine - it will be either defaults or user-set config data, so it's fine to push this entry to the end of the normalized list with its computed data cleared.